core/trino-main/src/main/java/io/trino/server/GracefulShutdownCoordinatorHandler.java

```diff
@@ -0,0 +1,194 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeoutException;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.concurrent.Threads.threadsNamed;
+import static java.lang.Thread.currentThread;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadExecutor;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public class GracefulShutdownCoordinatorHandler
+        implements GracefulShutdownHandler
+{
+    private static final Logger log = Logger.get(GracefulShutdownCoordinatorHandler.class);
+    private static final Duration LIFECYCLE_STOP_TIMEOUT = new Duration(30, SECONDS);
+
+    private final ScheduledExecutorService shutdownHandler = newSingleThreadScheduledExecutor(threadsNamed("shutdown-handler-%s"));
+    private final ExecutorService lifeCycleStopper = newSingleThreadExecutor(threadsNamed("lifecycle-stopper-%s"));
+    private final LifeCycleManager lifeCycleManager;
+    private final StartupStatus status;
+    private final QueryManager queryManager;
+    private final ShutdownAction shutdownAction;
+    private final Duration gracePeriod;
+    private final CoordinatorShutdownStrategy strategy;
+
+    @GuardedBy("this")
+    private boolean shutdownRequested;
+
+    @Inject
+    public GracefulShutdownCoordinatorHandler(
+            StartupStatus status,
+            QueryManager queryManager,
+            ServerConfig serverConfig,
+            ShutdownAction shutdownAction,
+            LifeCycleManager lifeCycleManager)
+    {
+        this.status = requireNonNull(status, "status is null");
+        this.queryManager = requireNonNull(queryManager, "queryManager is null");
+        this.shutdownAction = requireNonNull(shutdownAction, "shutdownAction is null");
+        this.lifeCycleManager = requireNonNull(lifeCycleManager, "lifeCycleManager is null");
+        this.gracePeriod = requireNonNull(serverConfig, "serverConfig is null").getGracePeriod();
+        this.strategy = serverConfig.getShutdownStrategy();
+    }
+
+    @Override
+    public synchronized void requestShutdown()
+    {
+        log.info("Coordinator shutdown requested");
+        if (shutdownRequested) {
+            return;
+        }
+        shutdownRequested = true;
+
+        // mark coordinator as shutting down - new query submissions will be rejected
+        status.shutdownStarted();
+        log.info("Coordinator is now rejecting new query submissions");
+
+        // wait for a grace period to start the shutdown sequence
+        shutdownHandler.schedule(this::shutdown, gracePeriod.toMillis(), MILLISECONDS);
```

Comment:
I've followed the same logic for worker shut down sequence. Maybe it's not needed for coordinator since we've marked as shutting down which will reject new query submissions